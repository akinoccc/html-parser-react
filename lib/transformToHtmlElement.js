/**
 * @name: AmsChan
 * @param {string} htmlstr
 * @param {object} data
 * @return {HTMLElement} dom[]
 */
export default (htmlstr, data = {}) => {
  const dataKeys = htmlstr.match(/\{.*\}/g) || [];
  dataKeys.forEach(target => {
    const key = target.replace(/[\{\}]/g, '');
    htmlstr = htmlstr.replace(target, data[key]);
  });
  const dom = document.createElement('div');
  dom.innerHTML = htmlstr;
  return dom.childNodes;
};
