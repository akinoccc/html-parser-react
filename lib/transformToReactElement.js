import React from 'react';
/**
 * @name: AmsChan
 * @param {HTMLElement} node
 * @param {JSX.Element[]} nodeTree
 * @param {object} option
 * @return {void}
 */
const _parseNode = (node, nodeTree, option) => {
  let type = node.nodeName.toLowerCase();
  const config = {};
  if (type.match('text')) {
    nodeTree.push(node.nodeValue);
    return;
  } else {
    const attrNames = Array.from(node.attributes || []);
    attrNames.forEach(attr => {
      let { nodeName, value } = attr;
      if (nodeName === 'style') {
        let styleObj = {};
        value.split(';').map(item => {
          if (item.trim().length === 0) {
            return;
          }
          let [key, v] = item.trim().split(':');
          const splitIndex = key.indexOf('-');
          if (splitIndex !== -1) {
            key = key.replace(/\-./, key[splitIndex + 1].toUpperCase());
          }
          let parserV;
          if (v.match('.')) {
            parserV = parseFloat(v);
          } else {
            parserV = parseInt(v);
          }
          if (isNaN(parserV)) {
            styleObj[key] = v;
          } else {
            styleObj[key] = parserV;
          }
        });
        value = styleObj;
      } else {
        const splitIndex = nodeName.indexOf('on');
        if (splitIndex !== -1) {
          let strArr = nodeName.split('');
          strArr[splitIndex + 2] = strArr[splitIndex + 2].toUpperCase();
          nodeName = strArr.join('');
          value = new Function(value);
        }
      }
      config[nodeName] = value;
    });
  }
  config['key'] = Date.now() + Math.random();
  nodeTree.push(React.createElement(type, config, type === 'img' ? null : []));
  transformToReactElement(
    node.childNodes,
    option,
    nodeTree[nodeTree.length - 1].props.children
  );
};

/**
 * @name: AmsChan
 * @param {HTMLElement[]} nodes
 * @param {object} option
 * @param {JSX.Element[]} nodeTree
 * @return {JSX.Element[]} nodeTree
 */
const transformToReactElement = (nodes, option = {}, nodeTree = []) => {
  const nodeArr = Array.from(nodes);
  nodeArr.forEach((node, index) => {
    if (option.transform) {
      const ret = option.transform(node, index);
      if (ret === null) {
        return;
      } else if (ret === undefined) {
        _parseNode(node, nodeTree, option);
      } else {
        nodeTree.push(ret);
      }
    } else {
      _parseNode(node, nodeTree, option);
    }
  });
  return nodeTree;
};
export default transformToReactElement;
