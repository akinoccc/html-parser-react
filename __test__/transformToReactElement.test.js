import transformToHtmlElement from '../lib/transformToHtmlElement';
import transformToReactElement from '../lib/transformToReactElement';
import htmlData from './testData/html.data';
import propsData from './testData/props.data';

describe('ReactParser', () => {
  it('test single html dom', () => {
    const htmlElements = transformToHtmlElement(htmlData.singleHtmlDom);
    const reactElements = transformToReactElement(htmlElements);
    expect(reactElements).toHaveLength(1);
    expect(reactElements[0].props.children).toHaveLength(1);
    expect(reactElements[0].props.children[0]).toBe('single test');
    expect(reactElements[0].type).toBe('div');
    expect(reactElements).toMatchSnapshot();
  });

  it('test mutiple html dom', () => {
    const htmlElements = transformToHtmlElement(htmlData.mutipleHtmlDom);
    const reactElements = transformToReactElement(htmlElements);
    expect(reactElements).toHaveLength(2);
    expect(reactElements[0].type).toBe('div');
    expect(reactElements[1].type).toBe('div');
    expect(reactElements[0].props.children).toHaveLength(1);
    expect(reactElements[1].props.children).toHaveLength(1);
    expect(reactElements[0].props.children[0].type).toBe('text');
    expect(reactElements[1].props.children[0].type).toBe('text');
    expect(reactElements[0].props.children[0].props.children[0]).toBe('first');
    expect(reactElements[1].props.children[0].props.children[0]).toBe('second');
    expect(reactElements).toMatchSnapshot();
  });

  it('test html attrs', () => {
    const htmlElements = transformToHtmlElement(htmlData.attributes);
    const reactElements = transformToReactElement(htmlElements);
    expect(reactElements).toHaveLength(1);
    expect(reactElements[0].props.id).toBe('test');
    expect(reactElements[0].props.src).toBe('null');
    expect(reactElements[0].props.children).toEqual(null);
    expect(reactElements[0].type).toBe('img');
    expect(reactElements).toMatchSnapshot();
  });

  it('test binding function', () => {
    const htmlElements = transformToHtmlElement(htmlData.bindFunctions);
    const reactElements = transformToReactElement(htmlElements);
    expect(reactElements).toHaveLength(1);
    expect(reactElements[0].props.children).toHaveLength(0);
    expect(typeof reactElements[0].props.onClick).toBe('function');
    expect(reactElements[0].props.onClick()).toBe(1);
    expect(reactElements[0].type).toBe('button');
    expect(reactElements).toMatchSnapshot();
  });

  it('test style', () => {
    const htmlElements = transformToHtmlElement(htmlData.style);
    const reactElements = transformToReactElement(htmlElements);
    expect(reactElements).toHaveLength(1);
    expect(reactElements[0].props.children).toHaveLength(0);
    expect(reactElements[0].props.style).toEqual({
      weight: 10,
      height: 5.5,
      marginLeft: 10,
      color: '#ccc',
      fontWeight: 700
    });
    expect(reactElements[0].type).toBe('div');
    expect(reactElements).toMatchSnapshot();
  });

  it('test render data', () => {
    const htmlElements = transformToHtmlElement(
      htmlData.renderData,
      propsData.data
    );
    const reactElements = transformToReactElement(htmlElements);
    expect(reactElements).toHaveLength(1);
    expect(reactElements[0].props.children).toHaveLength(2);
    expect(reactElements[0].props.children[1].props.children).toHaveLength(1);
    expect(reactElements[0].type).toBe('div');
    expect(reactElements[0].props.children[0]).toBe('this is data1');
    expect(reactElements[0].props.children[1].type).toBe('span');
    expect(reactElements[0].props.children[1].props.children[0]).toBe(
      'this is data2'
    );
    expect(reactElements).toMatchSnapshot();
  });

  it('transform function return undefine', () => {
    const htmlElements = transformToHtmlElement(htmlData.singleHtmlDom);
    const reactElements = transformToReactElement(
      htmlElements,
      propsData.option1
    );
    expect(reactElements).toHaveLength(1);
    expect(reactElements[0].props.children).toHaveLength(1);
    expect(reactElements[0].props.children[0]).toBe('single test');
    expect(reactElements[0].type).toBe('div');
    expect(reactElements).toMatchSnapshot();
  });

  it('transform function return null', () => {
    const htmlElements = transformToHtmlElement(htmlData.singleHtmlDom);
    const reactElements = transformToReactElement(
      htmlElements,
      propsData.option2
    );
    expect(reactElements).toHaveLength(0);
    expect(reactElements).toMatchSnapshot();
  });

  it('transform function return react element', () => {
    const htmlElements = transformToHtmlElement(htmlData.singleHtmlDom);
    const reactElements = transformToReactElement(
      htmlElements,
      propsData.option3
    );
    expect(reactElements).toHaveLength(1);
    expect(reactElements[0].props.children).toBe('transform');
    expect(reactElements[0].type).toBe('div');
    expect(reactElements).toMatchSnapshot();
  });
});
