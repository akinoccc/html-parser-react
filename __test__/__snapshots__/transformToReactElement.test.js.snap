// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactParser test binding function 1`] = `
Array [
  <button
    onClick={[Function]}
  />,
]
`;

exports[`ReactParser test html attrs 1`] = `
Array [
  <img
    id="test"
    src="null"
  />,
]
`;

exports[`ReactParser test mutiple html dom 1`] = `
Array [
  <div>
    <text>
      first
    </text>
  </div>,
  <div>
    <text>
      second
    </text>
  </div>,
]
`;

exports[`ReactParser test render data 1`] = `
Array [
  <div>
    this is data1
    <span>
      this is data2
    </span>
  </div>,
]
`;

exports[`ReactParser test single html dom 1`] = `
Array [
  <div>
    single test
  </div>,
]
`;

exports[`ReactParser test style 1`] = `
Array [
  <div
    style={
      Object {
        "color": "#ccc",
        "fontWeight": 700,
        "height": 5.5,
        "marginLeft": 10,
        "weight": 10,
      }
    }
  />,
]
`;

exports[`ReactParser transform function return null 1`] = `Array []`;

exports[`ReactParser transform function return react element 1`] = `
Array [
  <div>
    transform
  </div>,
]
`;

exports[`ReactParser transform function return undefine 1`] = `
Array [
  <div>
    single test
  </div>,
]
`;
